llm/bedrock/src/client.rs

```diff
@@ -0,0 +1,193 @@
+                    .await
+                    .map_err(|e| from_converse_stream_sdk_error(model_id, e));
+
+                trace!("Creating AWS Bedrock event stream");
+                match response {
+                    Ok(response) => BedrockChatStream::new(response.stream),
+                    Err(error) => BedrockChatStream::failed(error),
+                }
+            }
+        }
+    }
+
+    fn init_converse(&self, input: conversions::BedrockInput) -> ConverseFluentBuilder {
+        self.client
+            .converse()
+            .model_id(input.model_id)
+            .set_system(Some(input.system_instructions))
+            .set_messages(Some(input.messages))
+            .inference_config(input.inference_configuration)
+            .set_tool_config(input.tools)
+            .additional_model_request_fields(input.additional_fields)
+    }
+
+    fn init_converse_stream(
+        &self,
+        input: conversions::BedrockInput,
+    ) -> ConverseStreamFluentBuilder {
+        self.client
+            .converse_stream()
+            .model_id(input.model_id)
+            .set_system(Some(input.system_instructions))
+            .set_messages(Some(input.messages))
+            .inference_config(input.inference_configuration)
+            .set_tool_config(input.tools)
+            .additional_model_request_fields(input.additional_fields)
+    }
+}
+
+#[derive(Debug)]
+pub struct BedrockEnvironment {
+    access_key_id: String,
+    region: String,
+    secret_access_key: String,
+    session_token: Option<String>,
+}
+
+impl BedrockEnvironment {
+    pub fn load_from_env() -> Result<Self, llm::Error> {
+        Ok(Self {
+            access_key_id: get_config_key("AWS_ACCESS_KEY_ID")?,
+            region: get_config_key("AWS_REGION")?,
+            secret_access_key: get_config_key("AWS_SECRET_ACCESS_KEY")?,
+            session_token: get_config_key_or_none("AWS_SESSION_TOKEN"),
+        })
+    }
+
+    fn aws_region(&self) -> region::Region {
+        region::Region::new(self.region.clone())
+    }
+
+    fn aws_credentials(&self) -> bedrock::config::Credentials {
+        bedrock::config::Credentials::new(
+            self.access_key_id.clone(),
+            self.secret_access_key.clone(),
+            self.session_token.clone(),
+            None,
+            "llm-bedrock",
+        )
+    }
+}
+
+#[derive(Debug, Clone)]
+struct WasiSleep {
+    reactor: wasi_async_runtime::Reactor,
+}
+
+impl WasiSleep {
+    fn new(reactor: wasi_async_runtime::Reactor) -> Self {
+        Self { reactor }
+    }
+}
+
+impl AsyncSleep for WasiSleep {
+    fn sleep(&self, duration: std::time::Duration) -> Sleep {
+        let reactor = self.reactor.clone();
+        let fut = async move {
+            let nanos = duration.as_nanos() as u64;
+            let pollable = monotonic_clock::subscribe_duration(nanos);
+
+            reactor.wait_for(pollable).await;
+        };
+        Sleep::new(Box::pin(UnsafeFuture::new(fut)))
```

Comment:
This doesn't have to be done in this pr, but @kmatasfp pointed out that there is actually an even better runtime to use https://github.com/yoshuawuyts/wstd (basically a much more developed version of the wasi-async-runtime you are already using)

One noticeable improvement is that there is much better ways to thread the reactor through in order to convert Pollables to futures like here: https://github.com/yoshuawuyts/wstd/blob/main/src/runtime/reactor.rs#L110

Reply:
I would love to do it in this PR. let me make the changes.

Reply:
Done now!